# Data Exploration

In [1]:
import pandas as pd

## Movies Dataframe
The Movies dataframe contains detailed information about over 940,000 films, including their unique IDs, titles, release dates, poster taglines, short descriptions, durations in minutes, and their ratings.

In [2]:
movies_df = pd.read_csv('csv/movies.csv')

In [3]:
movies_df.shape

(941597, 7)

This dataset consists of 941,597 rows and 7 columns, providing comprehensive details about a vast collection of films.

In [4]:
movies_df.head()

id                               name    date  \
0  1000001                             Barbie  2023.0   
1  1000002                           Parasite  2019.0   
2  1000003  Everything Everywhere All at Once  2022.0   
3  1000004                         Fight Club  1999.0   
4  1000005                         La La Land  2016.0   

                                            tagline  \
0                  She's everything. He's just Ken.   
1                       Act like you own the place.   
2  The universe is so much bigger than you realize.   
3                           Mischief. Mayhem. Soap.   
4                    Here's to the fools who dream.   

                                         description  minute  rating  
0  Barbie and Ken are having the time of their li...   114.0    3.86  
1  All unemployed, Ki-taek's family takes peculia...   133.0    4.56  
2  An aging Chinese immigrant is swept up in an i...   140.0    4.30  
3  A ticking-time-bomb insomniac and a slippery s...   139.0    4.27  
4  Mia, an aspiring actress, serves lattes to mov...   129.0    4.09

In [5]:
movies_df.dtypes

id               int64
name            object
date           float64
tagline         object
description     object
minute         float64
rating         float64
dtype: object

The date column is currently stored as a float, which is not suitable for representing release dates; We will convert it to a proper date format for accurate analysis.

In [107]:
movies_df['date'] = pd.to_datetime(movies_df['date'])


In [ ]:

movies_df['name'] = movies_df['name'].astype('string')
movies_df['tagline'] = movies_df['tagline'].astype('string')
movies_df['description'] = movies_df['description'].astype('string')


In [7]:
movies_df.dtypes

id                      int64
name                   object
date           datetime64[ns]
tagline                object
description            object
minute                float64
rating                float64
dtype: object

### Analisi dei valori nulli

In [87]:
movies_df.isna().sum()

id                  0
name               10
date            91913
tagline        802210
description    160812
minute         181570
rating         850598
dtype: int64

##TODO LEVARE QUESTI FILM DALLA LISTA, PRIMA PERO VERIFICARE SE NONE E' SALVABILE
ho provato a pulire al massimo. qui io non hop piu trovato nulla da pulire _ yas

In [88]:
movies_df[movies_df['name'].isna()]
#rimuovo le righe con name nullo
movies_df = movies_df.dropna(subset=['name'])
#rimuovo le righe con nome strano o inutile 
#(film con il punto come nome ".")
movies_df = movies_df[movies_df['name'] != '.']
#rimuovo le apastrofi nei nomi che iniziano e finiscono con ' " '
movies_df['name'] = movies_df['name'].str.replace(r'^"|"$', '')
#rimuovo i film con nome uguale a "film"
movies_df = movies_df[movies_df['name'] != 'film']
#rimuovo i film con nome uguale a "movie"
movies_df = movies_df[movies_df['name'] != 'movie']
#rimuovo i film che hanno duplicati di nome 
movies_df = movies_df.drop_duplicates(subset='name')

In [11]:
movies_df.duplicated().sum()

np.int64(0)

In [12]:
movies_df['id'].duplicated().any()

np.False_

In [25]:
# Trova i film con nomi duplicati
movies_df['name'].duplicated().any()

np.True_

In [34]:
duplicati = movies_df[movies_df['name'].duplicated(keep=False)]

duplicati = duplicati.sort_values(by='name')

duplicati

id                                               name  \
812528  1812529                                             "Born"   
812712  1812713                                             "Born"   
926405  1926406  "How Will I Ever Get Out of This Labyrinth of ...   
582688  1582689  "How Will I Ever Get Out of This Labyrinth of ...   
755047  1755048                  "O Jehovah, . . . I Trust in You"   
...         ...                                                ...   
676309  1676310                                          정브르의 동물일기   
884959  1884960                                           조폭의 브이로그   
848930  1848931                                           조폭의 브이로그   
782298  1782299                                            코리안 커넥션   
871369  1871370                                            코리안 커넥션   

                                date                          tagline  \
812528 1970-01-01 00:00:00.000002023                              NaN   
812712 1970-01-01 00:00:00.000002023                              NaN   
926405                           NaT                              NaN   
582688 1970-01-01 00:00:00.000002023                              NaN   
755047 1970-01-01 00:00:00.000002016  O Jehovah, . . . I Trust in You   
...                              ...                              ...   
676309 1970-01-01 00:00:00.000002021                              NaN   
884959 1970-01-01 00:00:00.000002023                              NaN   
848930 1970-01-01 00:00:00.000002023                              NaN   
782298 1970-01-01 00:00:00.000001990                              NaN   
871369                           NaT                              NaN   

                                              description  minute  rating  
812528  Two teenagers sat and complained about coming ...    34.0     NaN  
812712  Two teenagers sat and complained about coming ...    34.0     NaN  
926405                                                NaN     NaN     NaN  
582688  Film student fails to complete her senior proj...    18.0     NaN  
755047  See how Hezekiah overcomes pressure from all s...    52.0     NaN  
...                                                   ...     ...     ...  
676309                                                NaN     NaN     NaN  
884959                                                NaN    57.0     NaN  
848930                                                NaN     NaN     NaN  
782298                                                NaN   117.0     NaN  
871369                                                NaN   117.0     NaN  

[201437 rows x 7 columns]

In [91]:
movies_df[(movies_df['rating'] > 5) | (movies_df['rating'] < 0)]

Empty DataFrame
Columns: [id, name, date, tagline, description, minute, rating]
Index: []

## Countries Dataframe
This dataset contains over 693,000 entries with the columns 'id', representing the film ID, and 'country', indicating one of the originating countries, with the possibility of multiple countries per film.

In [92]:
countries_df = pd.read_csv('../csv/countries.csv')

In [93]:
countries_df.shape

(693476, 2)

This dataset consists of 693,476 rows and 2 columns, providing detailed information about the countries associated with each film.

In [94]:
countries_df.head()

id      country
0  1000001           UK
1  1000001          USA
2  1000002  South Korea
3  1000003          USA
4  1000004      Germany

In [104]:
countries_df['country'] = countries_df['country'].astype('string')

In [105]:
# Verifica il tipo di dati della colonna 'country'
print(countries_df['country'].dtypes)
countries_df.dtypes

string


id                  int64
country    string[python]
dtype: object

In [20]:
countries_df.isna().sum()

id         0
country    0
dtype: int64

There are no null values in the dataset.

In [21]:
countries_df.duplicated().sum()

0

There are no duplicate entries in the dataset.

## Actors Dataframe
This dataset contains over 5.7 million entries with the columns 'id', representing the film ID, 'name', indicating the name of the actor, and 'role', specifying the character played.

In [22]:
actors_df = pd.read_csv('../csv/actors.csv')

In [23]:
actors_df.shape

(5798450, 3)

This dataset consists of 5,798,450 rows and 3 columns, providing detailed information about the actors, their roles, and the films they appear in.

In [24]:
actors_df.head()

id               name    role
0  1000001      Margot Robbie  Barbie
1  1000001       Ryan Gosling     Ken
2  1000001    America Ferrera  Gloria
3  1000001  Ariana Greenblatt   Sasha
4  1000001           Issa Rae  Barbie

In [108]:
actors_df['name'] = actors_df['name'].astype('string')
actors_df['role'] = actors_df['role'].astype('string')

actors_df.dtypes

id               int64
name    string[python]
role    string[python]
dtype: object

In [26]:
countries_df.isna().sum()

id         0
country    0
dtype: int64

There are no null values in the dataset.

In [27]:
countries_df.duplicated().sum()

0

There are no duplicate entries in the dataset.

## Crew Dataframe
This dataset contains over 4.7 million entries with the columns 'id', representing the film ID, 'role', indicating the role (e.g., director), and 'name', specifying the name of the person.

In [28]:
crew_df = pd.read_csv('../csv/crew.csv')

In [29]:
crew_df.shape

(4720183, 3)

This dataset consists of 4,720,183 rows and 3 columns, providing detailed information about the crew members, their roles, and the films they are associated with.

In [30]:
crew_df.head()

id      role            name
0  1000001  Director    Greta Gerwig
1  1000001  Producer    Tom Ackerley
2  1000001  Producer   Margot Robbie
3  1000001  Producer  Robbie Brenner
4  1000001  Producer    David Heyman

In [31]:

crew_df['role'] = crew_df['role'].astype('string')
crew_df['name'] = crew_df['name'].astype('string')

crew_df.dtypes

id       int64
role    object
name    object
dtype: object

In [32]:
crew_df.isna().sum()

id      0
role    0
name    1
dtype: int64

In [33]:
crew_df[crew_df['name'].isnull()]

id    role name
4562126  1859397  Writer  NaN

A null value was found in the 'name' column, but it will be retained as it may represent a missing or incomplete entry that is important for the context of the dataset.

In [34]:
countries_df.duplicated().sum()

0

There are no duplicate entries in the dataset.

## Posters Dataframe
This dataset contains over 940,000 entries with the columns 'id', representing the film ID, and 'link', providing the link to the original film posters.

In [35]:
posters_df = pd.read_csv('../csv/posters.csv')

In [36]:
posters_df.shape

(941597, 2)

This dataset consists of 941,597 rows and 2 columns, providing links to the original film posters associated with each film.

In [37]:
posters_df.head()

id                                               link
0  1000001  https://a.ltrbxd.com/resized/film-poster/2/7/7...
1  1000002  https://a.ltrbxd.com/resized/film-poster/4/2/6...
2  1000003  https://a.ltrbxd.com/resized/film-poster/4/7/4...
3  1000004  https://a.ltrbxd.com/resized/film-poster/5/1/5...
4  1000005  https://a.ltrbxd.com/resized/film-poster/2/4/0...

In [38]:
posters_df['link'] = posters_df['link'].astype('string')

posters_df.dtypes

id       int64
link    object
dtype: object

In [39]:
posters_df.isna().sum()

id           0
link    180712
dtype: int64

In [40]:
posters_df.duplicated().sum()

0

## Releases Dataframe
This dataset contains over 13 million entries with the columns 'id', representing the film ID, 'country', indicating the country of release, 'date', showing the release date, 'type', specifying the release type (e.g., Theatrical, Digital), and 'rating', representing the rating received in that country (e.g., PG, etc.).

In [41]:
releases_df = pd.read_csv('../csv/releases.csv')

In [42]:
releases_df.shape

(1332782, 5)

This dataset consists of 1,332,782 rows and 5 columns, providing detailed information about the film releases, including the country, date, type, and rating in each country.

In [43]:
releases_df.head()

id    country        date        type rating
0  1000001    Andorra  2023-07-21  Theatrical    NaN
1  1000001  Argentina  2023-07-20  Theatrical    ATP
2  1000001  Australia  2023-07-19  Theatrical     PG
3  1000001  Australia  2023-10-01     Digital     PG
4  1000001    Austria  2023-07-20  Theatrical    NaN

In [109]:
releases_df['country'] = releases_df['country'].astype('string')
releases_df['type'] = releases_df['type'].astype('string')
releases_df['rating'] = releases_df['rating'].astype('string')

releases_df.dtypes

id                  int64
country    string[python]
date       datetime64[ns]
type       string[python]
rating     string[python]
dtype: object

The date column is currently stored as an object, which is not suitable for representing release dates; We will convert it to a proper date format for accurate analysis.

In [110]:
releases_df['date'] = pd.to_datetime(releases_df['date'])
releases_df.dtypes

id                  int64
country    string[python]
date       datetime64[ns]
type       string[python]
rating     string[python]
dtype: object

In [46]:
releases_df.dtypes

id                  int64
country            object
date       datetime64[ns]
type               object
rating             object
dtype: object

In [47]:
releases_df.isna().sum()

id              0
country         0
date            0
type            0
rating     998802
dtype: int64

There are many null values in the 'rating' column, but no action will be taken as they will be left as is, since they may represent cases where a rating was not assigned in certain countries.

In [48]:
releases_df.duplicated().sum()

0

## Oscar Awards Dataframe
This dataset contains over 10,000 entries about Oscar candidates, with the columns 'year_film', representing the year the film was issued, 'year_ceremony', indicating the year of the ceremony, 'category', specifying the category (e.g., actor), 'name', showing the name of the person, 'film', listing the title of the film, and 'winner', indicating whether they won or not.

In [49]:
oscar_awards_df = pd.read_csv('../csv/the_oscar_awards.csv')

In [50]:
oscar_awards_df.shape

(10889, 7)

This dataset consists of 10,889 rows and 7 columns, providing detailed information about Oscar candidates, including the year of the film, ceremony, category, person, film title, and whether the candidate won.

In [51]:
oscar_awards_df.head()

year_film  year_ceremony  ceremony category                 name  \
0       1927           1928         1    ACTOR  Richard Barthelmess   
1       1927           1928         1    ACTOR        Emil Jannings   
2       1927           1928         1  ACTRESS       Louise Dresser   
3       1927           1928         1  ACTRESS         Janet Gaynor   
4       1927           1928         1  ACTRESS       Gloria Swanson   

               film  winner  
0         The Noose   False  
1  The Last Command    True  
2   A Ship Comes In   False  
3        7th Heaven    True  
4    Sadie Thompson   False

In [52]:
oscar_awards_df['category'] = oscar_awards_df['category'].astype('string')
oscar_awards_df['name'] = oscar_awards_df['name'].astype('string')
oscar_awards_df['film'] = oscar_awards_df['film'].astype('string')

oscar_awards_df.dtypes

year_film         int64
year_ceremony     int64
ceremony          int64
category         object
name             object
film             object
winner             bool
dtype: object

In [53]:
oscar_awards_df.isna().sum()

year_film          0
year_ceremony      0
ceremony           0
category           0
name               5
film             319
winner             0
dtype: int64

In [54]:
oscar_awards_df[oscar_awards_df['name'].isnull()]

year_film  year_ceremony  ceremony                          category  \
10513       2020           2021        93  JEAN HERSHOLT HUMANITARIAN AWARD   
10514       2020           2021        93  JEAN HERSHOLT HUMANITARIAN AWARD   
10635       2021           2022        94  JEAN HERSHOLT HUMANITARIAN AWARD   
10759       2022           2023        95  JEAN HERSHOLT HUMANITARIAN AWARD   
10885       2023           2024        96  JEAN HERSHOLT HUMANITARIAN AWARD   

      name film  winner  
10513  NaN  NaN    True  
10514  NaN  NaN    True  
10635  NaN  NaN    True  
10759  NaN  NaN    True  
10885  NaN  NaN    True

In [55]:
oscar_awards_df[oscar_awards_df['film'].isnull()]

year_film  year_ceremony  ceremony                          category  \
16          1927           1928         1               ENGINEERING EFFECTS   
18          1927           1928         1               ENGINEERING EFFECTS   
31          1927           1928         1           WRITING (Title Writing)   
32          1927           1928         1           WRITING (Title Writing)   
33          1927           1928         1                     SPECIAL AWARD   
...          ...            ...       ...                               ...   
10763       2022           2023        95            GORDON E. SAWYER AWARD   
10885       2023           2024        96  JEAN HERSHOLT HUMANITARIAN AWARD   
10886       2023           2024        96                    HONORARY AWARD   
10887       2023           2024        96                    HONORARY AWARD   
10888       2023           2024        96                    HONORARY AWARD   

                                                    name film  winner  
16                                        Ralph Hammeras  NaN   False  
18                                      Nugent Slaughter  NaN   False  
31                                        Joseph Farnham  NaN    True  
32                                    George Marion, Jr.  NaN   False  
33                                          Warner Bros.  NaN    True  
...                                                  ...  ...     ...  
10763                                          Iain Neil  NaN    True  
10885                                                NaN  NaN    True  
10886  To Angela Bassett, who has inspired audiences ...  NaN    True  
10887  To Mel Brooks, for his comedic brilliance, pro...  NaN    True  
10888  To Carol Littleton, whose commitment to her cr...  NaN    True  

[319 rows x 7 columns]

## Rotten Tomatoes Reviews Dataframe
This dataset contains over 1.1 million reviews from Rotten Tomatoes, with the columns 'rotten_tomatoes_link', providing the URL of the Rotten Tomatoes page, 'movie_title', indicating the title of the movie, 'critic_name', showing the name of the critic, 'top_critic', indicating if the critic is a top critic, 'publisher_name', specifying where the review was published, 'review_type', categorizing the type of review, 'review_score', representing the score given in the review, 'review_date', showing the date of the review, and 'review_content', containing the text of the actual review.

In [56]:
rotten_tomatoes_df = pd.read_csv('../csv/rotten_tomatoes_reviews.csv')

In [57]:
rotten_tomatoes_df.shape

(1129887, 9)

This dataset consists of 1,129,887 rows and 9 columns, providing detailed information about Rotten Tomatoes reviews, including the review URL, movie title, critic details, review score, date, and content.

In [58]:
rotten_tomatoes_df.head()

rotten_tomatoes_link                                        movie_title  \
0            m/0814255  Percy Jackson & the Olympians: The Lightning T...   
1            m/0814255  Percy Jackson & the Olympians: The Lightning T...   
2            m/0814255  Percy Jackson & the Olympians: The Lightning T...   
3            m/0814255  Percy Jackson & the Olympians: The Lightning T...   
4            m/0814255  Percy Jackson & the Olympians: The Lightning T...   

       critic_name  top_critic           publisher_name review_type  \
0  Andrew L. Urban       False           Urban Cinefile       Fresh   
1    Louise Keller       False           Urban Cinefile       Fresh   
2              NaN       False      FILMINK (Australia)       Fresh   
3     Ben McEachen       False  Sunday Mail (Australia)       Fresh   
4      Ethan Alter        True       Hollywood Reporter      Rotten   

  review_score review_date                                     review_content  
0          NaN  2010-02-06  A fantasy adventure that fuses Greek mythology...  
1          NaN  2010-02-06  Uma Thurman as Medusa, the gorgon with a coiff...  
2          NaN  2010-02-09  With a top-notch cast and dazzling special eff...  
3        3.5/5  2010-02-09  Whether audiences will get behind The Lightnin...  
4          NaN  2010-02-10  What's really lacking in The Lightning Thief i...

In [59]:
rotten_tomatoes_df.dtypes

rotten_tomatoes_link    object
movie_title             object
critic_name             object
top_critic                bool
publisher_name          object
review_type             object
review_score            object
review_date             object
review_content          object
dtype: object

The review_date column is currently stored as an object, which is not suitable for representing review dates; We will convert it to a proper date format for accurate analysis.

In [111]:
rotten_tomatoes_df['review_date'] = pd.to_datetime(rotten_tomatoes_df['review_date'])


rotten_tomatoes_df['movie_title'] = rotten_tomatoes_df['movie_title'].astype('string')
rotten_tomatoes_df['critic_name'] = rotten_tomatoes_df['critic_name'].astype('string')
rotten_tomatoes_df['publisher_name'] = rotten_tomatoes_df['publisher_name'].astype('string')
rotten_tomatoes_df['review_type'] = rotten_tomatoes_df['review_type'].astype('string')
rotten_tomatoes_df['review_score'] = rotten_tomatoes_df['review_score'].astype('string')
rotten_tomatoes_df['review_content'] = rotten_tomatoes_df['review_content'].astype('string')

rotten_tomatoes_df.dtypes

rotten_tomatoes_link            object
movie_title             string[python]
critic_name             string[python]
top_critic                        bool
publisher_name          string[python]
review_type             string[python]
review_score            string[python]
review_date             datetime64[ns]
review_content          string[python]
dtype: object

In [61]:
rotten_tomatoes_df.isna().sum()

rotten_tomatoes_link         0
movie_title                  0
critic_name              18521
top_critic                   0
publisher_name               0
review_type                  0
review_score            305902
review_date                  0
review_content           65778
dtype: int64

In [62]:
rotten_tomatoes_df[rotten_tomatoes_df['review_score'].isnull() &
                           rotten_tomatoes_df['review_content'].isnull() &
                           rotten_tomatoes_df['critic_name'].isnull()]

rotten_tomatoes_link                     movie_title  \
118910           m/alice_sweet_alice  Alice, Sweet Alice (Communion)   
219988              m/callas_forever                  Callas Forever   
323790    m/escape_to_witch_mountain        Escape to Witch Mountain   
349208                    m/flawless                        Flawless   
493769                m/kings_ransom                   King's Ransom   
619413                 m/nurse_betty                     Nurse Betty   
751877  m/saddest_music_in_the_world  The Saddest Music in the World   
751879  m/saddest_music_in_the_world  The Saddest Music in the World   

       critic_name  top_critic                publisher_name review_type  \
118910         NaN       False                         Film4      Rotten   
219988         NaN        True    Denver Rocky Mountain News      Rotten   
323790         NaN        True                Chicago Reader       Fresh   
349208         NaN       False                     E! Online      Rotten   
493769         NaN       False                 Hollywood.com       Fresh   
619413         NaN        True  Atlanta Journal-Constitution       Fresh   
751877         NaN       False             Premiere Magazine      Rotten   
751879         NaN       False             Premiere Magazine      Rotten   

       review_score review_date review_content  
118910          NaN  2003-05-24            NaN  
219988          NaN  2004-12-17            NaN  
323790          NaN  2000-01-01            NaN  
349208          NaN  2000-01-01            NaN  
493769          NaN  2005-04-23            NaN  
619413          NaN  2000-01-01            NaN  
751877          NaN  2004-07-03            NaN  
751879          NaN  2004-07-03            NaN

In [63]:
# Trova tutti i titoli di film unici nel dataset di Rotten Tomatoes
unique_rotten_tomatoes_movies = rotten_tomatoes_df['movie_title'].unique()

# Trova i titoli di film unici nel dataset dei film
unique_movies = movies_df['name'].unique()

# Converti tutti i titoli di film in stringhe e poi in minuscolo per il confronto
unique_rotten_tomatoes_movies = [str(title).lower() for title in unique_rotten_tomatoes_movies]
unique_movies = [str(title).lower() for title in unique_movies]

# Trova i film che sono presenti in entrambi i dataset
common_movies = set(unique_rotten_tomatoes_movies).intersection(set(unique_movies))

# Stampa il numero di film trovati
print(f"Numero di film unici in Rotten Tomatoes: {len(unique_rotten_tomatoes_movies)}")
print(f"Numero di film unici nel dataset dei film: {len(unique_movies)}")
print(f"Numero di film comuni: {len(common_movies)}")

Numero di film unici in Rotten Tomatoes: 17100
Numero di film unici nel dataset dei film: 795422
Numero di film comuni: 15106


## Tentativo di capirci qualcosa sui dati rovinati 

In [64]:
# Remove invalid movies
invalid_movies = movies_df[movies_df['name'].isna() | (movies_df['name'] == '.')]
movies_df = movies_df.drop(invalid_movies.index)

In [65]:
# Remove invalid Oscar entries
invalid_oscar = oscar_awards_df[oscar_awards_df['name'].isna() & oscar_awards_df['film'].isna()]
oscar_awards_df = oscar_awards_df.drop(invalid_oscar.index)

In [66]:
# Remove invalid reviews
invalid_reviews = rotten_tomatoes_df[rotten_tomatoes_df['review_score'].isna() & rotten_tomatoes_df['review_content'].isna() & rotten_tomatoes_df['critic_name'].isna()]
rotten_tomatoes_df = rotten_tomatoes_df.drop(invalid_reviews.index)

Yassek: ok, io onesto toglierei queste cose, non vorrei comparissero nella nostra app

In [69]:
# Standardize names for matching
def standardize_name(name):
    return name.lower().replace(' ', '_').replace('"', '')

# Fill NaN values with an empty string before applying the function
movies_df['name'] = movies_df['name'].fillna('').apply(standardize_name)
oscar_awards_df['film'] = oscar_awards_df['film'].fillna('').apply(standardize_name)
rotten_tomatoes_df['movie_title'] = rotten_tomatoes_df['movie_title'].fillna('').apply(standardize_name)

In [ ]:
# Match movies with reviews
unique_rotten_tomatoes_movies = rotten_tomatoes_df['movie_title'].unique()
unique_movies = movies_df['name'].unique()
common_movies = set(unique_rotten_tomatoes_movies).intersection(set(unique_movies))
print(f"Numero di film unici in Rotten Tomatoes: {len(unique_rotten_tomatoes_movies)}")
print(f"Numero di film unici nel dataset dei film: {len(unique_movies)}")
print(f"Numero di film comuni: {len(common_movies)}")